### ISSUE TITLE
Add optional support for external $ref in CLI

### ISSUE THREAD

**Issue by philsturgeon** [stoplightio/spectral#134]
### **I'm submitting a...**
  - [ ] bug report
  - [x] feature request

### What is the current behavior?

The default behavior of Spectral v2.0 is to operate only on the original description object for 95% of the rules, and the uses the `resolvedTarget` in a few rules which specifically request that resolved version of the object (like oasOpsParams). 

Spectral CLI uses [json-ref-parser](https://github.com/stoplightio/json-ref-resolver) to created that resolvedTarget, and by default json-ref-parser only processes internal `$ref`. This is a safe default, because $ref can be potentially used maliciously if you are working with description files you don't control.  

This is unexpected for people transitioning from Speccy, which did external by default. See #108.

### What is the expected behavior?

Maybe we could have a new CLI option for the `lint` command, maybe `--external`.

The main `Spectral` constructor takes an object with a `resolver` key, which defaults to a stock standard json-ref-parser instance. The lint command should take advantage of this and change the resolver based on CLI options. 

```
if (flags.external) {
    const spectral = new Spectral(resolver: <fancy-resolver-however-that-works>);
}
```

Or something like that.

Thoughts @tbarn @marbemac @MikeRalphson?

**Comment by marbemac** [comment]
Ok so `json-ref-resolver` already supports resolving remote references, we just need to tell it how to. Adding remote resolution to spectral CLI would be as simple as updating the cli lint command to pass in a resolver instance that has readers for http and file, like so:

https://github.com/stoplightio/spectral/blob/master/src/cli/commands/lint.ts#L81

```ts
import { Resolver } from "@stoplight/json-ref-resolver";

// whatever http library, here using axios as example
import * as axios from "axios";

// only possible in node.. but ok since we're in cli right now
import * as fs from "fs";

const httpReader = {
  async read(ref: uri.URI) {
    return axios({
      method: "get",
      url: String(ref)
    });
  }
}

// create our resolver instance
const resolver = new Resolver({
  // readers can do anything, so long as they define an async read function that resolves to a value
  readers: {
    // these readers will be invoked for refs with the http(s)? protocol
    https: httpReader,
    http: httpReader,

    // this reader will be invoked for refs with the file protocol
    file: {
      async read(ref: uri.URI) {
        return fs.read(String(ref));
      }
    }
  }
});

async function lint(name: string, flags: any, command: Lint) {
  // ... rest

  const spectral = new Spectral({ resolver });

  // ... rest
}
```

Can find the example of this in the json-ref-resolver readme: https://github.com/stoplightio/json-ref-resolver#example-resolving-remote-references-with-readers

That should be enough to get both http and file resolution working in spectral cli. The reason not do this in the default resolver that spectral main class uses (https://github.com/stoplightio/spectral/blob/master/src/spectral.ts#L2) is because as you can see we’re importing fs and axios. `fs` won’t work in the browser for most folks, hence the idea of just making a couple out of the box resolvers available for import at specific paths in the lib. Those paths should not be exported from main entry point. something like:

**spectral/src/resolvers/internal.ts**

```ts
import { Resolver } from "@stoplight/json-ref-resolver";

// just resolves internal #/definitions/user type $refs
export const localResolver = new Resolver();
```

**spectral/src/resolvers/http.ts**

```ts
import { Resolver } from "@stoplight/json-ref-resolver";
import * as axios from "axios";

const httpReader = {
  async read(ref) {
    return axios({
      method: "get",
      url: String(ref)
    });
  }
}

// resolves http and https $refs, and internal $refs
export const httpResolver = new Resolver({
  readers: {
    https: httpReader,
    http: httpReader,
  }
});
```

**spectral/src/resolvers/httpAndFile.ts**

```ts
import { Resolver } from "@stoplight/json-ref-resolver";
import * as axios from "axios";
import * as fs from "fs";

const httpReader = {
  async read(ref) {
    return axios({
      method: "get",
      url: String(ref)
    });
  }
}

// resolves files, http and https $refs, and internal $refs
export const httpAndFileResolver = new Resolver({
  readers: {
    https: httpReader,
    http: httpReader,
    file: {
      async read(ref) {
        return fs.read(String(ref));
      }
    }
  }
});
```

Spectral library user (and cli actually would use the httpAndFile one) could then use these like this:

```ts
import { Spectral } from '@stoplight/spectral';
import { httpResolver } from '@stoplight/spectral/resolvers/http';

const spectral = new Spectral({ resolver: httpResolver });
```

This makes loading axios and fs into the code path optional - only occurs if you import one of those resolvers. End user can of course also provide their own custom resolver if they want to add readers for `mongo://` or other exotic protocols (json-ref-resolver supports this).

**Comment by philsturgeon** [comment]
Just a little update, this one fell through the cracks a bit, so I have cranked up the priority on it a little and hopefully it'll get sorted out next sprint.

**Comment by blazeworx** [comment]
Thanks Phil! Really looking forward to switching over to spectral as soon as this one is in :) 👍

**Comment by marbemac** [comment]
Another update, this one is confirmed for our sprint starting this Thursday - will be in soon!
